### PR TITLE
Feature/optimalcontrol

### PR DESCRIPTION
**Pull request by S-Dafarra**
First skeleton of a optimal control library (very skinny for the time being). 

This branch is on top of improvements/spline, for which a PR is already opened https://github.com/robotology/idyntree/pull/398

### CODE REVIEW

**Comment by traversaro** [comment]
If we plan to keep this library "unstable" (i.e. have the API of the library to change without deprecations) we should add this classes to iDynTreeExperimental, as in https://github.com/robotology/idyntree/blob/a982b71e0adfd7f2f32eb152bde70922fbb24389/src/estimation/include/iDynTree/Estimation/BerdySparseMAPSolver.h#L39 . 
Check also the codacy comment: https://github.com/robotology/idyntree/pull/399#discussion_r155264923 .

**Comment by S-Dafarra** [comment]
Regarding the Codacy comment https://github.com/robotology/idyntree/pull/399#issuecomment-349669884  I don't understand completely why it complains. Isn't it fine to use the operator ``++`` with an iterator?

Regarding the ``iDynTreeExperimental`` tag, is there a way to do it automatically for the whole directory?

**Comment by traversaro** [comment]
There are also a few warnings that may be worth checking: https://travis-ci.org/robotology/idyntree/jobs/312456782#L7303 .

**Comment by traversaro** [comment]
Fixed some CMake problems with the tests in https://github.com/robotology/idyntree/pull/399/commits/53db94c7663633075d568f9f5ed68520c06216bd .

File: src/optimalcontrol/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+set(libraryname idyntree-optimalcontrol)
+
+
+idyntree_enable_cxx11()
+
+option(BUILD_SHARED_LIBS "Build libraries as shared" TRUE)
+
+# Dependencies
+# IPOPT
+# SUNDIALS (?)
```
Comment (traversaro) [inline src/optimalcontrol/CMakeLists.txt]: I guess in the end you went for implementing your own integrator, we can remove this comments.

File: src/optimalcontrol/tests/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+#add_test(NAME OptimalControlTest
+#        COMMAND ${PROJECT_NAME})
+
+add_executable(ConstraintsGroupTest ConstraintsGroupTest.cpp)
+
+target_link_libraries(ConstraintsGroupTest PRIVATE idyntree-optimalcontrol ${iDynTree_LIBRARIES})
+
+
+add_test(NAME ConstraintsGroupTest
```
Comment (traversaro) [inline src/optimalcontrol/tests/CMakeLists.txt]: Again a low priority comment, but we should follow the same scheme of the tests of the other components and also enabled Valgrind-based tests.

**Comment by traversaro** [comment]
> Regarding the iDynTreeExperimental tag, is there a way to do it automatically for the whole directory? 

Unfortunately no, but the basic idea is that every class belongs to a doxygen group, so eventually that part could be switched to a `\ingroup iDynTreeOptimalControl`.

**Comment by traversaro** [comment]
> Regarding the Codacy comment #399 (comment) I don't understand completely why it complains. Isn't it fine to use the operator ++ with an iterator?**

For objects that are not integers or similarly simple objects, it is better (even is typically the difference is negligible) to use the **prefix** ++ in place of the **postfix** ++, see http://www.idinews.com/peeves/prefixIncr.html .

**Comment by traversaro** [comment]
The only failure is related to https://github.com/robotology/yarp/issues/1486 , merging. The rest of open issues are stored in https://github.com/robotology/idyntree/issues/400 .

**Pull request merged**